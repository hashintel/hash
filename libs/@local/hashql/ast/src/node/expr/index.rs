use hashql_core::{heap, span::SpanId};

use super::Expr;
use crate::node::id::NodeId;

/// An indexing expression in the HashQL Abstract Syntax Tree.
///
/// Represents accessing an element from a collection by its index or key.
/// Indexing works with various collection types including lists, tuples,
/// dictionaries, structs, and other indexed data structures.
///
/// The index expression can be any expression that evaluates to a valid
/// index for the collection type - typically an integer for lists or a value of appropriate key
/// type for dictionaries.
///
/// Tuples and structs can be indexed using this expression, but only if the type system can prove
/// that the index is within bounds.
///
/// # Examples
///
/// ## J-Expr
///
/// ```json
/// ["[]", "items", 0]
/// ["items[0]"]
/// ["matrix[i][j]"]
/// ```
///
/// ## Documentation Format
///
/// ```text
/// items[0]
/// matrix[i][j]
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct IndexExpr<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub value: heap::Box<'heap, Expr<'heap>>,
    pub index: heap::Box<'heap, Expr<'heap>>,
}
