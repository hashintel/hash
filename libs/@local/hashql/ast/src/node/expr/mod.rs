//! Expression nodes for the HashQL Abstract Syntax Tree.
//!
//! This module defines all expression types used in HashQL, a side-effect free, purely
//! functional programming language for querying bi-temporal graphs. The expressions are
//! organized into a hierarchy, with [`Expr`] at the root containing various specific
//! expression kinds represented by the [`ExprKind`] enum.
//!
//! # Special Forms
//!
//! Several expression kinds (`Let`, `Use`, `Input`, `Closure`, `If`, `Field`, `Index`) are
//! implemented as "special forms". Initially, these are parsed as ordinary function calls into the
//! AST. During a subsequent AST transformation phase, these function calls are recognized by their
//! names and expanded into their corresponding specialized AST nodes. For example, a `let`
//! function call is transformed into a dedicated [`LetExpr`] structure in the AST after parsing.
//!
//! # Memory Management
//!
//! The AST is designed to work with arena allocation through the `'heap` lifetime,
//! which references the memory region where nodes are allocated. This approach
//! improves performance by minimizing allocations and deallocations during parsing
//! and evaluation.
//!
//! Unlike other steps, in which the AST is interned in a "real" arena allocator, the AST is using
//! `Box` to manage memory instead. This is important as some of the steps in the AST require that
//! we do not deduplicate nodes yet. This makes each node in the tree larger, but also has the
//! side-effect of easier access of the tree.
//!
//! # Node Identification and Source Location
//!
//! Each expression node has a unique identifier (`id`) that can be used to track the node through
//! various processing stages. Additionally, nodes carry a span identifier (`span`) that points to
//! the location of the expression in the source code, facilitating error reporting and debugging.
//!
//! # Frontend
//!
//! While the AST itself is frontend-independent, the examples shown in this documentation use
//! `JExpr`, which is a JSON-based syntax for HashQL. Other frontends may be added in the future,
//! all mapping to this same core AST structure.
pub mod call;
pub mod closure;
pub mod dict;
pub mod field;
pub mod r#if;
pub mod index;
pub mod input;
pub mod is;
pub mod r#let;
pub mod list;
pub mod literal;
pub mod newtype;
pub mod r#struct;
pub mod tuple;
pub mod r#type;
pub mod r#use;

use hashql_core::span::SpanId;

pub use self::{
    call::CallExpr, closure::ClosureExpr, dict::DictExpr, field::FieldExpr, r#if::IfExpr,
    index::IndexExpr, input::InputExpr, is::IsExpr, r#let::LetExpr, list::ListExpr,
    literal::LiteralExpr, newtype::NewTypeExpr, r#struct::StructExpr, tuple::TupleExpr,
    r#type::TypeExpr, r#use::UseExpr,
};
use super::{id::NodeId, path::Path};

/// Represents the different kinds of expressions in HashQL.
///
/// This enum defines all the various expression types that can appear in a HashQL program.
/// Each variant corresponds to a specific language construct and contains the detailed
/// representation of that construct in the AST.
///
/// The examples below demonstrate the `JExpr` syntax (JSON-based frontend), as well as a fictional
/// "documentation syntax" (used for readability) for each expression kind. Remember that these are
/// just frontend representations - the AST itself is independent of any particular syntax.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum ExprKind<'heap> {
    /// A function call expression.
    ///
    /// Represents an expression that invokes a function with arguments.
    /// Arguments can be positional or labeled (keyword arguments).
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// // Positional arguments only
    /// ["add", 1, 2]
    ///
    /// // Labeled arguments only
    /// ["add", {":name": "Alice"}, {":age": 30}]
    ///
    /// // Mix of positional and labeled arguments
    /// ["format", 42, {":width": 10}, {":precision": 2}]
    ///
    /// // Advanced function call
    /// ["let", "func", ["if", "condition", "fn1", "fn2"],
    ///     ["func", "value"]
    /// ]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// add(1, 2)
    /// add(alice: "Alice", age: 30)
    /// format("Hello, {name}!", name: "Alice")
    ///
    /// let func = if condition then fn1 else fn2 in
    /// func(value)
    /// ```
    Call(CallExpr<'heap>),

    /// A struct instantiation expression.
    ///
    /// Creates a new anonymous struct instance with the specified fields.
    /// In HashQL, structs are heterogeneous collections of named fields with associated values.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// {"#struct": {"field1": "value1", "field2": "value2"}}
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// (field1: value1, field2: value2)
    /// ```
    Struct(StructExpr<'heap>),

    /// A dictionary expression.
    ///
    /// Creates a dictionary (map) with key-value pairs. Unlike structs,
    /// dictionaries have dynamically computed keys and support lookups,
    /// but their values must be of a homogeneous type.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// {"#dict": {"key1": "value1", "key2": "value2"}}
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// {"key1": value1, "key2": value2}
    /// ```
    Dict(DictExpr<'heap>),

    /// A tuple expression.
    ///
    /// Creates a tuple with multiple values. Tuples in HashQL are
    /// fixed-size heterogeneous collections accessed by position.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// {"#tuple": ["value1", "value2", "value3"]}
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// (value1, value2, value3)
    /// ```
    Tuple(TupleExpr<'heap>),

    /// A list expression.
    ///
    /// Creates a list containing multiple elements. Lists in HashQL are
    /// homogeneous collections of values that can be accessed by index.
    ///
    /// # Examples
    ///
    /// ```json
    /// {"#list": ["value1", "value2", "value3"]}
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// [value1, value2, value3]
    /// ```
    List(ListExpr<'heap>),

    /// A literal expression.
    ///
    /// Represents a constant value like a number, string, boolean, or null.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// {"#literal": 123}
    /// {"#literal": "hello"}
    /// {"#literal": true}
    /// {"#literal": null}
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// 123
    /// "hello"
    /// true
    /// null
    /// ```
    Literal(LiteralExpr<'heap>),

    /// A variable reference expression.
    ///
    /// References a variable or qualified path.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// "variable_name"
    /// "graph::user::name"
    /// "graph::user<A, B>"
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// variable_name
    /// graph::user::name
    /// graph::user<A, B>
    /// ```
    Path(Path<'heap>),

    /// A `let` binding expression (special form).
    ///
    /// Binds a value to a name within a scope. This is expanded from a function call
    /// during AST transformation. In HashQL, `let` expressions create lexically
    /// scoped bindings visible within the body expression.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["let", "x", 42, <body>]
    /// ["let", "x", "Int", 42, <body>]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// let x = 42 in <body>
    /// let x: Int = 42 in <body>
    /// ```
    Let(LetExpr<'heap>),

    /// A type definition expression (special form).
    ///
    /// Defines a type alias within a scope. This is expanded from a function call
    /// during AST transformation. In HashQL, type expressions allow creating named
    /// references to complex types.
    /// These types use structural typing, meaning two types are considered equivalent
    /// if they have the same structure, regardless of their names.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["type", "UserId", "String", <body>]
    /// ["type", "Point", {"#type": {"x": "Float", "y": "Float"}}, <body>]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// type UserId = String in <body>
    /// type Point = {x: Float, y: Float} in <body>
    /// ```
    Type(TypeExpr<'heap>),

    /// A new type definition expression (special form).
    ///
    /// Creates a new distinct type based on an existing type. This is expanded
    /// from a function call during AST transformation. Unlike type aliases (created with
    /// the `type` expression), new types are not interchangeable with their underlying
    /// type.
    ///
    /// When defining a new type, a constructor function of the same name is generated automatically
    /// and brought into scope.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["newtype", "UserId", "String", <body>]
    /// ["newtype", "Coordinates", {"#type": {"lat": "Float", "lng": "Float"}}, <body>]
    ///
    /// ["newtype", "AccountId", "String"
    ///     ["AccountId", {"#literal": "1234"}]
    /// ]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// newtype UserId = String in <body>
    /// newtype Coordinates = {lat: Float, lng: Float} in <body>
    ///
    /// newtype AccountId = String in
    ///     AccountId("1234")
    /// ```
    NewType(NewTypeExpr<'heap>),

    /// A module import expression (special form).
    ///
    /// Imports symbols from another module into the current scope. This is expanded
    /// from a function call during AST transformation. HashQL supports selective
    /// imports with optional renaming.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["use", "path::to::module", {"#struct": {"item1": "_", "original": "renamed"}}, <body>]
    /// ["use", "path::to::module", {"#tuple": ["item1", "original"]}, <body>]
    /// ["use", "path::to::module", "*", <body>]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// use path::to::module::{item1, original as renamed} in <body>
    /// use path::to::module::{item1, original} in <body>
    /// use path::to::module::* in <body>
    /// ```
    Use(UseExpr<'heap>),

    /// An input parameter declaration (special form).
    ///
    /// Declares an input parameter for a function or query. This is expanded
    /// from a function call during AST transformation. In HashQL, inputs
    /// represent the external values that can be provided to a query.
    ///
    /// In effectful languages these are often called the context or requirements.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["input", "limit", "Int"]
    /// ["input", "limit", "Int", 10]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// input(limit, Int)
    /// input(limit, Int, 10)
    /// ```
    Input(InputExpr<'heap>),

    /// A closure expression (special form).
    ///
    /// Defines an anonymous function. This is expanded from a function call
    /// during AST transformation. In HashQL, closures can capture variables
    /// from their surrounding scope.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["fn", [], {"#struct": {"x": "Int"}}, "Int", ["*", "x", 2]]
    ///
    /// ["fn", {"#tuple": ["T"]}, {"#struct": {"x": "T", "y": "T"}}, "T", ["*", "x", "y"]]
    /// ["fn", {"#struct": {"T": "Int"}}, {"#struct": {"x": "T", "y": "T"}}, "_", ["*", "x", "y"]]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// fn(x: Int): Int => *(x, 2)
    ///
    /// fn<T>(x: T, y: T): T => *(x, y)
    /// fn<T: Int>(x: T, y: T): T => *(x, y)
    /// ```
    Closure(ClosureExpr<'heap>),

    /// A conditional expression (special form).
    ///
    /// Evaluates a condition and returns one of two expressions based on the result.
    /// This is expanded from a function call during AST transformation.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["if", [">", "x", 0], "positive", "non_positive"]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// if x > 0
    ///     then positive
    ///     else non_positive
    /// ```
    If(IfExpr<'heap>),

    /// A field access expression (special form).
    ///
    /// Accesses a field of a struct or tuple. This is expanded from a function call
    /// during AST transformation.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// [".", "user", "name"]
    /// ["user.name"]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// user.name
    /// ```
    Field(FieldExpr<'heap>),

    /// An indexing expression (special form).
    ///
    /// Accesses an element of a collection by index. This is expanded from a function call
    /// during AST transformation. Indexing works with lists, tuples, structs, dictionaries and
    /// other indexed collections in HashQL.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["[]", "items", 0]
    /// ["items[0]"]
    /// ["matrix[i][j]"]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// items[0]
    /// matrix[i][j]
    /// ```
    Index(IndexExpr<'heap>),

    /// A type assertion expression (special form).
    ///
    /// Checks at compile time whether a value conforms to a specified type.
    /// This is useful for type narrowing and ensuring type safety in patterns
    /// where the compiler needs additional type information.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// ["is", "value", "String"]
    /// ["is", ["get", "data", "field"], {"#type": {"name": "String", "age": "Int"}}]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// value is String
    /// get(data, field) is {name: String, age: Int}
    /// ```
    Is(IsExpr<'heap>),

    /// The underscore expression, used as a placeholder in various contexts.
    ///
    /// In HashQL, the underscore serves as a context-dependent placeholder that can indicate
    /// type inference, function generic bounds, or same-name imports.
    ///
    /// # Examples
    ///
    /// ## J-Expr
    ///
    /// ```json
    /// // Type inference in function return type
    /// ["fn", {"#tuple": []}, {"#struct": {"name": "String"}}, "_", ["body"]]
    ///
    /// // No bounds for a generic
    /// ["fn", {"#struct": {"A": "_"}}, {"#struct": {"name": "A"}}, "Integer", ["body"]]
    ///
    /// // Same-name import in struct pattern
    /// ["use", "module", {"#struct": {"name": "_"}}, "body"]
    ///
    /// // Same-name import in tuple pattern
    /// ["use", "module", {"#tuple": ["name"]}, "body"]
    /// ```
    ///
    /// ## Documentation Format
    ///
    /// ```text
    /// fn(): _ => body
    /// use module::{name: _} in body
    /// use module::{name} in body
    /// ```
    Underscore,

    /// A placeholder expression used exclusively during AST transformation phases.
    ///
    /// The `Dummy` variant serves as a temporary placeholder during the lowering process when an
    /// expression's `ExprKind` needs to be extracted or replaced. It allows the lowering pass to
    /// safely move out the contents of an expression without immediately providing a replacement.
    ///
    /// # Implementation Note
    ///
    /// This variant should never appear in a fully processed AST, as it's intended only as an
    /// intermediate state during transformation. Any `Dummy` nodes still present after lowering
    /// indicates an error in the transformation process, and will produce a compilation error on
    /// lowering into the HIR.
    Dummy,
}

/// An expression node in the HashQL Abstract Syntax Tree.
///
/// The `Expr` struct is the fundamental building block of the HashQL AST.
/// It represents any kind of expression in the language and can be composed
/// to form complex expressions. Every expression is allocated on a heap
/// referenced by the `'heap` lifetime.
///
/// Each expression has a unique identifier and a span that points to its
/// location in the source code, which are crucial for error reporting,
/// debugging, and tracking nodes through transformation phases.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Expr<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub kind: ExprKind<'heap>,
}

impl Expr<'_> {
    pub(crate) const fn dummy() -> Self {
        Self {
            id: NodeId::PLACEHOLDER,
            span: SpanId::SYNTHETIC,
            kind: ExprKind::Dummy,
        }
    }
}
