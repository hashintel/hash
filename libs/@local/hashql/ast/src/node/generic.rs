use hashql_core::{span::SpanId, symbol::Ident};

use super::{id::NodeId, r#type::Type};
use crate::heap;

/// A concrete type argument for a generic type.
///
/// Represents a specific type used to instantiate a generic type parameter.
/// In a generic type instantiation like `Vec<String>`, `String` would be
/// represented as a [`GenericArgument`].
///
/// # Examples
///
/// In a type reference like `HashMap<K, V>`, `K` and `V` are generic arguments.
/// In an instantiated type like `List<Int>`, `Int` is a generic argument.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct GenericArgument {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
}

/// A generic type parameter declaration.
///
/// Represents a type parameter in a generic type or function declaration.
/// Each parameter has a name and an optional type bound that constrains
/// what types can be used as arguments for this parameter.
///
/// # Examples
///
/// In a declaration like `fn<T>(x: T): T`, `T` is a generic parameter.
/// In a declaration like `fn<T: Number>(x: T): T`, `T` is a generic parameter
/// with a bound of `Number`.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct GenericParam<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
    pub bound: Option<heap::Box<'heap, Type<'heap>>>,
}

/// A collection of generic parameters for a type or function.
///
/// Represents the complete set of type parameters in a generic declaration,
/// such as the `<T, U>` in `fn<T, U>(x: T, y: U): U`.
///
/// # Examples
///
/// For a function like `fn<T, U: Comparable>(x: T, y: U): Bool`:
/// - The `Generics` would contain two `GenericParam` entries
/// - One for `T` with no bound
/// - One for `U` with a bound of `Comparable`
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Generics<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub params: heap::Box<'heap, [GenericParam<'heap>]>,
}
