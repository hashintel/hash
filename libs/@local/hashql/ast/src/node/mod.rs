//! Core Abstract Syntax Tree node types for HashQL.
//!
//! This module defines the fundamental building blocks of the HashQL Abstract
//! Syntax Tree (AST). The AST represents the structure of a HashQL program after
//! parsing, capturing all the language constructs in a hierarchical tree structure.
//!
//! ## Module Organization
//!
//! The node module is organized into several submodules:
//!
//! - [`expr`]: Expression nodes representing computations that produce values
//! - [`generic`]: Generic type parameter and argument definitions
//! - [`id`]: Node identifier types and utilities
//! - [`path`]: Path expressions for variable references and qualified names
//! - [`type`]: Type system representations including struct, tuple, and other types
//!
//! ## Memory Management
//!
//! All AST nodes use arena allocation through the `'heap` lifetime parameter,
//! which references the memory region where nodes are allocated. This approach
//! improves performance by minimizing allocations and simplifies memory management
//! during parsing and analysis.
//!
//! ## Node Identification
//!
//! Each node in the AST has a unique [`id::NodeId`] that can be used to track
//! the node through various processing stages. Nodes also include a span identifier that points to
//! the source location of the node.
//!
//! ## AST Structure
//!
//! The AST is structured as a tree with expressions ([`expr::Expr`]) at the core.
//! Other node types like types ([`type::Type`]) and paths ([`path::Path`]) are
//! used as components within expressions or on their own for declarations.
//!
//! The entire tree represents a complete HashQL program or query, ready for
//! analysis, transformation, and evaluation.
pub mod expr;
pub mod generic;
pub mod id;
pub mod path;
pub mod r#type;
