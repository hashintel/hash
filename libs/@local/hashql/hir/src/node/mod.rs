pub mod access;
pub mod branch;
pub mod call;
pub mod closure;
pub mod data;
pub mod graph;
pub mod input;
pub mod kind;
pub mod r#let;
pub mod operation;
pub mod variable;

use hashql_core::{id, span::SpanId};

use self::kind::NodeKind;

id::newtype!(
    /// A unique identifier for nodes in the HashQL High-Level Intermediate Representation (HIR).
    ///
    /// Each node in the HIR has a unique `NodeId` that can be used to track the node
    /// through various processing stages. These IDs are particularly important for
    /// error reporting, cross-referencing between nodes, and maintaining node identity
    /// during transformations.
    ///
    /// The value space is restricted to 0..=0xFFFF_FF00, reserving the last 256 for niches.
    /// As real pattern types are an experimental feature in Rust, these can currently only be
    /// used by directly modifying and accessing the `NodeId`'s internal value.
    pub struct HirId(u32 is 0..=0xFFFF_FF00)
);

impl HirId {
    /// A placeholder ID used during initial parsing.
    ///
    /// When parsing, we initially give all AST nodes this placeholder ID. After resolving special
    /// forms, the nodes are renumbered to have small indices.
    pub const PLACEHOLDER: Self = <Self as id::Id>::MAX;
}

/// A node in the HashQL High-Level Intermediate Representation (HIR).
///
/// The HIR is an optimized, more refined representation of the program derived from the AST.
/// While the AST closely mirrors the source syntax, the HIR represents a more semantically
/// meaningful structure that's better suited for type checking, optimization, and code generation.
///
/// Key differences between HIR and AST:
/// - Special forms (like `let`, `if`, etc.) are fully expanded in the HIR
/// - Syntactic sugar is desugared and normalized
/// - The HIR uses interned references for better memory efficiency
/// - Nodes are more compact and optimized for compiler operations
///
/// Each HIR node has:
/// - A unique identifier for tracking through compilation phases
/// - A span linking to its origin in source code for error reporting
/// - A specific kind that determines the node's semantics and structure
///
/// The HIR is designed to be immutable after construction and uses arena allocation
/// through the `'heap` lifetime for efficient memory management.
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct Node<'heap> {
    pub id: HirId,
    pub span: SpanId,

    pub kind: NodeKind<'heap>,
}
