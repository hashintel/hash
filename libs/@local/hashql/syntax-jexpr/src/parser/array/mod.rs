pub(crate) mod error;
pub(crate) mod visit;

use hashql_ast::node::{
    expr::{
        CallExpr, Expr, ExprKind,
        call::{Argument, LabeledArgument},
    },
    id::NodeId,
};

use self::{
    error::{empty, labeled_argument_invalid_identifier, labeled_argument_missing_prefix},
    visit::visit_array,
};
use super::{error::ParserDiagnostic, string::parse_ident_labelled_argument_from_string};
use crate::{
    error::ResultExt as _,
    lexer::{syntax_kind::SyntaxKind, token::Token, token_kind::TokenKind},
    parser::{
        error::ParserDiagnosticCategory, expr::parse_expr, object::visit::visit_object,
        state::ParserState,
    },
    span::Span,
};

// Peek twice, we know if it's a labeled argument if the first token is `{` and the second a
// string that starts with `:`.
fn parse_labelled_argument<'heap>(
    state: &mut ParserState<'heap, '_>,
) -> Result<Option<Vec<LabeledArgument<'heap>>>, ParserDiagnostic> {
    let Some(peek1) = state
        .peek()
        .change_category(ParserDiagnosticCategory::Lexer)?
    else {
        return Ok(None);
    };

    if peek1.kind.syntax() != SyntaxKind::LBrace {
        return Ok(None);
    }

    let Some(peek2) = state
        .peek2()
        .change_category(ParserDiagnosticCategory::Lexer)?
    else {
        return Ok(None);
    };

    let TokenKind::String(key) = &peek2.kind else {
        return Ok(None);
    };

    if !key.starts_with(':') {
        return Ok(None);
    }

    let token = state
        .advance(SyntaxKind::LBrace)
        .change_category(From::from)?;

    let mut labeled_arguments = Vec::new();

    visit_object(state, token, |state, key| {
        if !key.value.starts_with(':') {
            return Err(
                labeled_argument_missing_prefix(state.insert_range(key.span), key.value)
                    .map_category(From::from),
            );
        }

        let mut label_span = key.span;

        let key_span = state.insert_range(key.span);
        let key = match parse_ident_labelled_argument_from_string(state, key_span, &key.value) {
            Ok(value) => value,
            Err(error) => {
                return Err(
                    labeled_argument_invalid_identifier(state.spans(), key_span, error)
                        .map_category(From::from),
                );
            }
        };

        let value = parse_expr(state)?;
        label_span = label_span.cover(state.current_span());

        labeled_arguments.push(LabeledArgument {
            id: NodeId::PLACEHOLDER,
            span: state.insert_range(label_span),
            label: key,
            value: Argument {
                id: NodeId::PLACEHOLDER,
                span: value.span,
                value: state.heap().boxed(value),
            },
        });

        Ok(())
    })?;

    Ok(Some(labeled_arguments))
}

pub(crate) fn parse_array<'heap, 'source>(
    state: &mut ParserState<'heap, 'source>,
    token: Token<'source>,
) -> Result<Expr<'heap>, ParserDiagnostic> {
    debug_assert_eq!(token.kind.syntax(), SyntaxKind::LBracket);

    let mut function = None;
    let mut arguments = Vec::new();
    let mut labeled_arguments = Vec::new();

    let span = visit_array(state, token, |state| {
        match &mut function {
            Some(_) => {
                if let Some(labeled) = parse_labelled_argument(state)? {
                    labeled_arguments.extend(labeled);
                    return Ok(());
                }

                let expr = parse_expr(state)?;

                arguments.push(Argument {
                    id: NodeId::PLACEHOLDER,
                    span: expr.span,
                    value: state.heap().boxed(expr),
                });
            }
            function @ None => *function = Some(parse_expr(state)?),
        }

        Ok(())
    })
    .change_category(From::from)?;

    let span = state.insert_span(Span {
        range: span,
        pointer: Some(state.current_pointer()),
        parent_id: None,
    });

    let Some(function) = function else {
        return Err(empty(span).map_category(From::from));
    };

    let heap = state.heap();

    Ok(Expr {
        id: NodeId::PLACEHOLDER,
        span,
        kind: ExprKind::Call(CallExpr {
            id: NodeId::PLACEHOLDER,
            span,
            function: heap.boxed(function),
            arguments: heap.transfer_vec(arguments),
            labeled_arguments: heap.transfer_vec(labeled_arguments),
        }),
    })
}

#[cfg(test)]
mod tests {
    use hashql_ast::format::SyntaxDump as _;
    use insta::{assert_snapshot, with_settings};

    use crate::{
        lexer::syntax_kind::SyntaxKind,
        parser::{array::parse_array, test::bind_parser},
    };

    bind_parser!(fn run_array(parse_array, SyntaxKind::LBracket));

    #[test]
    fn parse_basic_function_call() {
        // ["add", 1, 2] - Basic function call structure
        let result = run_array(r##"["add", {"#literal": 1}, {"#literal": 2}]"##)
            .expect("should parse successfully");

        with_settings!({
            description => "Parses a basic function call with proper literal syntax"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_empty_array() {
        // [] - Empty array should error (as it has no function)
        let error = run_array("[]").expect_err("should fail on empty array");

        with_settings!({
            description => "Empty arrays are not valid J-Expr function calls"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_function_call_with_labeled_arguments() {
        // Function call with labeled arguments
        let result = run_array(
            r##"["greet", {":name": {"#literal": "Alice"}}, {":age": {"#literal": 30}}]"##,
        )
        .expect("should parse successfully");

        with_settings!({
            description => "Parses a function call with labeled arguments"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_function_call_mixed_arguments() {
        // Function call with both positional and labeled arguments
        let result =
            run_array(r##"["format", {"#literal": "template"}, {":value": {"#literal": 42}}]"##)
                .expect("should parse successfully");

        with_settings!({
            description => "Parses a function call with mixed positional and labeled arguments"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_nested_function_calls() {
        // Nested function calls
        let result = run_array(
            r##"["add", ["multiply", {"#literal": 2}, {"#literal": 3}], {"#literal": 5}]"##,
        )
        .expect("should parse successfully");

        with_settings!({
            description => "Parses nested function calls"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_labeled_argument_missing_prefix() {
        // Missing ":" prefix in labeled argument
        let error = run_array(r##"["greet", {"name": {"#literal": "Alice"}}]"##)
            .expect_err("should fail with missing prefix");

        with_settings!({
            description => "Labeled arguments must start with ':'"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_labeled_argument_invalid_identifier() {
        // Invalid identifier starting with digit
        let error = run_array(r##"["func", {":123": {"#literal": "value"}}]"##)
            .expect_err("should fail with invalid identifier");

        with_settings!({
            description => "Labeled arguments must have valid identifiers"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }

    #[test]
    fn parse_labeled_argument_with_complex_value() {
        // Labeled argument with a nested function call
        let result = run_array(r#"["func", {":config": ["getConfig"]}]"#)
            .expect("should parse successfully");

        with_settings!({
            description => "Parses a labeled argument with a complex value"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_multiline_array() {
        // Multiline function call with formatting
        let result = run_array(
            r##"[
            "let",
            {"#literal": "x"},
            {"#literal": 10},
            ["add", {"#literal": "x"}, {"#literal": 5}]
            ]"##,
        )
        .expect("should parse successfully");

        with_settings!({
            description => "Parses a multiline function call with proper formatting"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_labeled_argument_complex_nested() {
        // A labeled argument with a nested labeled argument
        let result =
            run_array(r##"["func", {":config": ["setup", {":nested": {"#literal": true}}]}]"##)
                .expect("should parse successfully");

        with_settings!({
            description => "Parses a labeled argument containing nested labeled arguments"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, result.dump, &result.input);
        });
    }

    #[test]
    fn parse_malformed_labeled_argument() {
        // Malformed labeled argument syntax with multiple keys
        let error = run_array(
            r##"["func", {":name": {"#literal": "value"}, "extra": {"#literal": true}}]"##,
        )
        .expect_err("should fail with multiple keys in a labeled argument object");

        with_settings!({
            description => "Malformed labeled argument with multiple keys should fail"
        }, {
            assert_snapshot!(insta::_macro_support::AutoName, error.diagnostic, &error.input);
        });
    }
}
