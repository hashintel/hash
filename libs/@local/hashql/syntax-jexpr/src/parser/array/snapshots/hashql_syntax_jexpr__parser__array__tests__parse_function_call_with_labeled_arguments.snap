---
source: libs/@local/hashql/syntax-jexpr/src/parser/array/mod.rs
description: Parses a function call with labeled arguments
expression: "[\"greet\", {\":name\": {\"#literal\": \"Alice\"}}, {\":age\": {\"#literal\": 30}}]"
---
Expr#4294967040@14
  ExprKind (Call)
    CallExpr#4294967040@14
      Expr#4294967040@3
        ExprKind (Path)
          Path#4294967040@3 (rooted: false)
            PathSegment#4294967040@2 (name: greet)
      LabeledArgument#4294967040@8 (label: name)
        Argument#4294967040@7
          Expr#4294967040@7
            ExprKind (Literal)
              LiteralExpr#4294967040@6
                LiteralKind (String)
                  StringLiteral (Alice)
      LabeledArgument#4294967040@13 (label: age)
        Argument#4294967040@12
          Expr#4294967040@12
            ExprKind (Literal)
              LiteralExpr#4294967040@11
                LiteralKind (Integer)
                  IntegerLiteral (30)
